### PR TITLE
fix(ivy): preventDefault when listener returns false

### LINKED ISSUE #22495: Render3 should support listener return value
Render3 needs to retain return values of event handlers so that preventDefault is called when appropriate in the DOM renderer.

https://github.com/angular/angular/blob/e95b61d42a3cacc8206417ef15387c865a059618/packages/platform-browser/src/dom/dom_renderer.ts#L52-L56

Currently for a click handler like this:

```html
<button (click)="onClick()"></button>
```

we generate:

```ts
E(0, 'button');
  L('click', function() { ctx.onClick(); } )  ;
e();
```

However, this loses the return value. The compiler could generate:

```ts
E(0, 'button');
  L('click', function() { return ctx.onClick(); } )  ;
e();
```

This is slightly more complicated for chains like:

```html
<button (click)="onClick();  otherFn();"></button>
```

Which becomes:

```ts
E(0, 'button');
  L('click', function() { ctx.onClick(); return ctx.otherFn(); } )  ;
e();
```

Downsides: more code size for each listener defined :-( 

Since this involves compiler changes, @chuckjaz will probably want to weigh in on what makes the most sense to generate.

cc @chuckjaz

### PR DESCRIPTION
**Pull request by kara**
This PR ensures that returning false from an event handler calls `preventDefault()`, as it does in current Angular.

Closes #22495

### CODE REVIEW

File: packages/core/test/render3/compiler_canonical/compiler_canonical_spec.ts [hunk 1]
```diff
@@ -58,6 +58,45 @@ describe('compiler specification', () => {
...
+        // NORMATIVE
+        static ngComponentDef = $r3$.ɵdefineComponent({
+          type: ListenerComp,
+          tag: 'listener-comp',
+          factory: function ListenerComp_Factory() { return new ListenerComp(); },
+          template: function ListenerComp_Template(ctx: $ListenerComp$, cm: $boolean$) {
+            if (cm) {
+              $r3$.ɵE(0, 'button');
+              $r3$.ɵL('click', function ListenerComp_click_Handler() { return ctx.onClick(); });
+              $r3$.ɵL('keypress', function ListenerComp_keypress_Handler($e$: $any$) {
```
Comment (chuckjaz) [inline packages/core/test/render3/compiler_canonical/compiler_canonical_spec.ts]: `$e$` should be `$event`. This name should be canonical as it is documented to obscure a field by the same name and you are required to use `this.$event` to access the field that is obscured.

**Review by chuckjaz — APPROVED** [review]
Minor nit.

**Review by mhevery — APPROVED** [review]

**Comment by kara** [comment]
[presubmit](http://test/OCL:187499044:BASE:187499111:1519929758279:3e7f2a88)

File: packages/core/src/render3/instructions.ts [hunk 2]
```diff
@@ -1437,10 +1439,28 @@ export function markDirtyIfOnPush(node: LElementNode): void {
...
+    return listenerFn(e);
+  };
+}
+
+/**
+ * Wraps an event listener so its host view and its ancestor views will be marked dirty
+ * whenever the event fires. Also wraps with preventDefault behavior.
+ */
+export function wrapListenerWithDirtyAndDefault(
+    view: LView, listenerFn: (e?: any) => any): EventListener {
```
Comment (vicb) [inline packages/core/src/render3/instructions.ts]: - should both the function return a consistent type (`EventListener`) ?
- should both the event listeners return a consistent value (`listenerFn(e)`) ?

Comment (kara) [reply packages/core/src/render3/instructions.ts]: The `EventListener` type returns void and the renderer2 function must relay the return value (it is wrapped in dom_renderer), so they cannot both be `EventListener`. Given that the `wrapListenerWithDirtyAndDefault` version is sent to `addEventListener`, I don't think it makes sense to pass it something that isn't type `EventListener`.

**Pull request closed without merging**